File: packages/plugins/http/Sources/TanoPluginHTTP/HTTPPlugin.swift
import Foundation
import TanoBridge

/// Native HTTP client plugin for Tano.
///
/// Uses URLSession to make HTTP requests directly from the native layer,
/// bypassing WebView CORS restrictions. Works on both iOS and macOS.
///
/// Supported methods: `request`.
public final class HTTPPlugin: TanoPlugin {

    // MARK: - TanoPlugin conformance

    public static let name = "http"
    public static let permissions: [String] = ["network"]

    public init() {}

    // MARK: - Routing

    public func handle(method: String, params: [String: Any]) async throws -> Any? {
        switch method {
        case "request":
            return try await request(params: params)
        default:
            throw HTTPPluginError.unknownMethod(method)
        }
    }

    // MARK: - request

    private func request(params: [String: Any]) async throws -> [String: Any] {
        guard let urlString = params["url"] as? String else {
            throw HTTPPluginError.missingParameter("url")
        }

        guard let url = URL(string: urlString) else {
            throw HTTPPluginError.invalidParameter("url is not a valid URL: \(urlString)")
        }

        let httpMethod = (params["method"] as? String)?.uppercased() ?? "GET"
        let headers = params["headers"] as? [String: String] ?? [:]
        let bodyString = params["body"] as? String

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = httpMethod

        for (key, value) in headers {
            urlRequest.setValue(value, forHTTPHeaderField: key)
        }

        if let bodyString = bodyString {
            urlRequest.httpBody = Data(bodyString.utf8)
        }

        let (data, response) = try await URLSession.shared.data(for: urlRequest)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw HTTPPluginError.invalidResponse
        }

        // Convert response headers to [String: String]
        var responseHeaders: [String: String] = [:]
        for (key, value) in httpResponse.allHeaderFields {
            if let k = key as? String, let v = value as? String {
                responseHeaders[k] = v
            }
        }

        let bodyText = String(data: data, encoding: .utf8) ?? data.base64EncodedString()

        return [
            "status": httpResponse.statusCode,
            "headers": responseHeaders,
            "body": bodyText,
        ]
    }
}

// MARK: - Errors

public enum HTTPPluginError: Error, LocalizedError {
    case unknownMethod(String)
    case missingParameter(String)
    case invalidParameter(String)
    case invalidResponse

    public var errorDescription: String? {
        switch self {
        case .unknownMethod(let m):
            return "Unknown HTTP plugin method: \(m)"
        case .missingParameter(let p):
            return "Missing required parameter: \(p)"
        case .invalidParameter(let p):
            return "Invalid parameter: \(p)"
        case .invalidResponse:
            return "Received a non-HTTP response"
        }
    }
}

File: packages/android/src/main/kotlin/dev/tano/plugins/HTTPPlugin.kt
package dev.tano.plugins

import dev.tano.bridge.TanoPlugin
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * Native HTTP client plugin for Tano on Android.
 *
 * Uses [HttpURLConnection] to make HTTP requests directly from the native
 * layer, bypassing WebView CORS restrictions. No external dependencies
 * (e.g. OkHttp) required.
 *
 * All network I/O runs on [Dispatchers.IO].
 *
 * Supported methods: `request`.
 *
 * Mirrors the iOS HTTPPlugin.
 */
class HTTPPlugin : TanoPlugin {

    override val name: String = "http"
    override val permissions: List<String> = listOf("network")

    // -- Routing --

    override suspend fun handle(method: String, params: Map<String, Any?>): Any? {
        return when (method) {
            "request" -> request(params)
            else -> throw IllegalArgumentException("Unknown HTTP plugin method: $method")
        }
    }

    // -- request --

    @Suppress("UNCHECKED_CAST")
    private suspend fun request(params: Map<String, Any?>): Map<String, Any?> =
        withContext(Dispatchers.IO) {
            val urlString = params["url"] as? String
                ?: throw IllegalArgumentException("Missing required parameter: url")

            val url = try {
                URL(urlString)
            } catch (e: Exception) {
                throw IllegalArgumentException("Invalid parameter: url is not a valid URL: $urlString")
            }

            val httpMethod = (params["method"] as? String)?.uppercase() ?: "GET"
            val headers = params["headers"] as? Map<String, String> ?: emptyMap()
            val bodyString = params["body"] as? String

            val connection = url.openConnection() as HttpURLConnection
            try {
                connection.requestMethod = httpMethod
                connection.connectTimeout = 30_000
                connection.readTimeout = 30_000

                for ((key, value) in headers) {
                    connection.setRequestProperty(key, value)
                }

                if (bodyString != null) {
                    connection.doOutput = true
                    connection.outputStream.use { output ->
                        output.write(bodyString.toByteArray(Charsets.UTF_8))
                    }
                }

                val statusCode = connection.responseCode

                // Read response headers
                val responseHeaders = mutableMapOf<String, String>()
                for (i in 0 until connection.headerFields.size) {
                    val key = connection.getHeaderFieldKey(i) ?: continue
                    val value = connection.getHeaderField(i) ?: continue
                    responseHeaders[key] = value
                }

                // Read response body
                val inputStream = try {
                    connection.inputStream
                } catch (_: Exception) {
                    connection.errorStream
                }

                val bodyText = if (inputStream != null) {
                    BufferedReader(InputStreamReader(inputStream, Charsets.UTF_8)).use { reader ->
                        reader.readText()
                    }
                } else {
                    ""
                }

                mapOf(
                    "status" to statusCode,
                    "headers" to responseHeaders,
                    "body" to bodyText
                )
            } finally {
                connection.disconnect()
            }
        }
}
